File: engine/measures.ts
// Measure-level similarity: weighted lexical DAX features + greedy one-to-one matching.
// Ported from semantic_sweep/measures.py (constants kept identical for parity).

import {
  type DaxFeatures,
  type Measure,
  type MeasureMatch,
  disjoint,
  intersection,
  round4,
  union,
} from "./types";

const AGGREGATORS = new Set([
  "sum", "average", "min", "max", "count", "counta", "countrows", "distinctcount",
  "sumx", "averagex", "minx", "maxx", "countx", "product", "productx", "median", "geomean",
]);

const FUNCTION_FAMILIES: Record<string, string> = {
  round: "rounding", roundup: "rounding", rounddown: "rounding", trunc: "rounding",
  int: "rounding", mround: "rounding", fixed: "rounding", ceiling: "rounding", floor: "rounding",
  sum: "additive", sumx: "additive",
  average: "mean", averagex: "mean",
  count: "counting", counta: "counting", countrows: "counting", countx: "counting",
  distinctcount: "counting",
};

const CONTEXT_FLAGS = new Set([
  "calculate", "calculatetable", "filter", "all", "allexcept", "allselected", "removefilters",
  "userelationship", "keepfilters", "sameperiodlastyear", "dateadd", "totalytd", "totalmtd",
  "totalqtd", "datesytd", "parallelperiod", "previousmonth", "previousyear", "datesinperiod",
]);

const GENERIC_NAMES = new Set([
  "total", "count", "sum", "amount", "value", "measure", "result", "kpi", "average", "max", "min",
]);

// Column/measure ref names common enough to recur across unrelated tables on any estate with a
// shared naming convention (e.g. every fact table has an [Amount], every dimension has a [Date]).
// A bare match on one of these must not, by itself, count as ref-backed strong-duplicate evidence
// (acc6) -- unlike GENERIC_NAMES above (measure-name weighting), this gates measure-level evidence.
const GENERIC_REF_NAMES = new Set([
  "id", "key", "code", "name", "date", "value", "amount", "total", "count", "sum", "status", "type",
  "description", "flag", "number", "quantity", "price", "region", "category", "year", "month", "day",
  "created", "modified", "updated",
]);

const COMPONENT_WEIGHTS: Record<string, number> = { refs: 0.45, functions: 0.3, flags: 0.15, operators: 0.1 };
const SKELETON_SCORE = 0.92;
// Same structural skeleton but the referenced names differ: score between SKELETON_FLOOR (pure shape
// collision, e.g. SUM(Sales[x]) vs SUM(HR[y])) and SKELETON_SCORE (renamed table, identical column),
// interpolated by how many referenced column/measure names coincide. Kept >= REVIEW_MEASURE so a
// fully-renamed clone still surfaces for review; the strong-duplicate gate is enforced separately
// via ref-backed evidence (see matchModelMeasures.strongMatched).
const SKELETON_FLOOR = 0.9;
const AGG_PENALTY = 0.6;

// A DAX string literal, escape-aware: an embedded literal `"` is written as a doubled `""` inside the
// string (e.g. `"He said ""hi"""`). A naive `"[^"]*"` stops at the first inner `"`, truncating the
// literal and leaking the remainder as if it were DAX code (dax-tokenizer-hardening).
const STRING_SRC = String.raw`"(?:[^"]|"")*"`;
const STRING_RE = new RegExp(STRING_SRC, "g");
// One pass, string-aware: tries a full string literal FIRST at each position, so a "//" or "/*" that
// appears INSIDE a string is consumed as part of the string and never misread as a comment start (the
// previous two-pass comment stripper had no string awareness at all).
const COMMENT_OR_STRING_RE = new RegExp(`${STRING_SRC}|/\\*[\\s\\S]*?\\*/|//[^\\n]*`, "g");

export function normalizeDax(dax: string): string {
  const text = dax.replace(COMMENT_OR_STRING_RE, (m) => (m.startsWith('"') ? m : " "));
  return text.toLowerCase().replace(/\s+/g, " ").trim();
}

function skeleton(norm: string): string {
  // `[\p{L}\p{N}_]` (+ `u` flag) rather than `\w`: JS's `\w` is ASCII-only, unlike Python's (Unicode-aware
  // by default), so a non-ASCII bare table qualifier (e.g. `Clientèle[Amount]`) would otherwise be split
  // mid-identifier here, diverging from the Python engine's parse of the identical DAX expression.
  let text = norm.replace(/(?:'[^']*'|[\p{L}\p{N}_]+)?\[[^\]]*\]/gu, "#r#");
  text = text.replace(STRING_RE, "#s#");
  text = text.replace(/\b\d+(?:\.\d+)?\b/g, "#n#");
  return text.replace(/\s+/g, "");
}

function matchAll(re: RegExp, text: string, group = 0): string[] {
  const out: string[] = [];
  let m: RegExpExecArray | null;
  const r = new RegExp(re.source, re.flags.includes("g") ? re.flags : re.flags + "g");
  while ((m = r.exec(text)) !== null) {
    out.push(m[group]);
    if (m.index === r.lastIndex) r.lastIndex++;
  }
  return out;
}

// "table.column" refs, captured only when a table qualifier (bare word or 'quoted name') immediately
// precedes the bracket in the source DAX. A bare [Column] with no qualifier (row-context reference or
// a measure call) contributes nothing here -- it stays in the unqualified `refs` set instead.
function extractQualifiedRefs(text: string): Set<string> {
  const out = new Set<string>();
  // Same Unicode-aware fix as skeleton() above: `[\p{L}\p{N}_]` + `u` flag instead of ASCII-only `\w`.
  const re = /(?:'([^']*)'|([\p{L}\p{N}_]+))?\[([^\]]+)\]/gu;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    const qualifier = m[1] ?? m[2];
    if (qualifier) out.add(`${qualifier}.${m[3]}`);
    if (m.index === re.lastIndex) re.lastIndex++;
  }
  return out;
}

export function extractFeatures(dax: string): DaxFeatures {
  const norm = normalizeDax(dax);
  const functions = new Set(matchAll(/\b([a-z][a-z0-9]*)\s*\(/g, norm, 1));
  // Neutralize string literals before extracting column/measure refs so a format string like
  // FORMAT(x, "[Red]0") does not leak a bogus "red" reference (acc3).
  const noStrings = norm.replace(STRING_RE, '""');
  return {
    norm,
    skeleton: skeleton(norm),
    functions,
    refs: new Set(matchAll(/\[([^\]]+)\]/g, noStrings, 1).map((s) => s.toLowerCase())),
    qualifiedRefs: extractQualifiedRefs(noStrings),
    aggregators: intersection(functions, AGGREGATORS),
    operators: new Set(matchAll(/[+\-*/&<>=]/g, norm)),
    flags: intersection(functions, CONTEXT_FLAGS),
  };
}

// Two matched measures are "ref-backed" (real evidence, not a coincidental shape collision) when:
// identical normalized DAX, OR an exact table-qualified column/measure match, OR a shared BARE name
// that is specific enough to not be a coincidence. A bare match on a generic name alone (e.g. two
// unrelated tables both having an [Amount] or [Date] column) is NOT accepted -- that previously let
// Sales[Amount] and Budget[Amount] manufacture strong-duplicate evidence on shared-schema estates
// (acc6). Renamed-table clones with specific shared column names (e.g. Sales[amt0]~Revenue[amt0])
// still count, preserving the acc5 renamed-clone heuristic.
function isRefBacked(a: DaxFeatures, b: DaxFeatures): boolean {
  if (a.norm === b.norm) return true;
  if (intersection(a.qualifiedRefs, b.qualifiedRefs).size > 0) return true;
  const shared = intersection(a.refs, b.refs);
  for (const name of shared) if (!GENERIC_REF_NAMES.has(name)) return true;
  return false;
}

function jac(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0;
  return intersection(a, b).size / union(a, b).size;
}

function functionSimilarity(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0;
  const exact = intersection(a, b);
  const famA = new Set<string>();
  const famB = new Set<string>();
  for (const f of a) if (!exact.has(f) && FUNCTION_FAMILIES[f]) famA.add(FUNCTION_FAMILIES[f]);
  for (const f of b) if (!exact.has(f) && FUNCTION_FAMILIES[f]) famB.add(FUNCTION_FAMILIES[f]);
  const soft = exact.size + 0.5 * intersection(famA, famB).size;
  return Math.min(1, soft / union(a, b).size);
}

export function measureSimilarity(a: DaxFeatures, b: DaxFeatures): number {
  // No DAX on either side (e.g. a locked-down / admin scan that withholds expressions) is NOT
  // evidence of a match — two shape-identical models must not read as exact clones (acc1).
  if (!a.norm && !b.norm) return 0;
  if (a.norm === b.norm) return 1;
  if (a.skeleton && a.skeleton === b.skeleton) {
    return round4(SKELETON_FLOOR + (SKELETON_SCORE - SKELETON_FLOOR) * jac(a.refs, b.refs));
  }
  const scores: Record<string, number> = {
    refs: jac(a.refs, b.refs),
    functions: functionSimilarity(a.functions, b.functions),
    flags: jac(a.flags, b.flags),
    operators: jac(a.operators, b.operators),
  };
  const present: Array<[string, boolean]> = [
    ["refs", a.refs.size > 0 || b.refs.size > 0],
    ["functions", a.functions.size > 0 || b.functions.size > 0],
    ["flags", a.flags.size > 0 || b.flags.size > 0],
    ["operators", a.operators.size > 0 || b.operators.size > 0],
  ];
  const active = present.filter(([, has]) => has).map(([k]) => k);
  if (active.length === 0) return 0;
  const wsum = active.reduce((s, k) => s + COMPONENT_WEIGHTS[k], 0);
  let base = active.reduce((s, k) => s + COMPONENT_WEIGHTS[k] * scores[k], 0) / wsum;
  if (
    a.aggregators.size > 0 &&
    b.aggregators.size > 0 &&
    disjoint(a.aggregators, b.aggregators) &&
    intersection(a.refs, b.refs).size > 0
  ) {
    base *= AGG_PENALTY;
  }
  return round4(Math.min(1, base));
}

function measureWeight(m: Measure, f: DaxFeatures): number {
  // Tokenize on word boundaries (not a single concatenated blob) so a multi-word name like "Total
  // Sales" is judged word-by-word against GENERIC_NAMES instead of vanishing into "totalsales",
  // which matches nothing and was silently never downweighted. Square the generic-word fraction
  // before applying the discount: a name that is ALL generic words (e.g. "Total", or "Total Count")
  // still gets the full 0.4 floor exactly as before, but a name that is only PARTLY generic (e.g.
  // "Total Sales") is discounted much more gently, since one specific word ("Sales") already carries
  // most of the discriminating signal -- informativeness doesn't fall off linearly with word count.
  const words = m.name.toLowerCase().split(/[^a-z0-9]+/).filter(Boolean);
  const genericFrac = words.length ? words.filter((w) => GENERIC_NAMES.has(w)).length / words.length : 0;
  const base = 1 - 0.6 * genericFrac * genericFrac;
  const complexity = 1 + 0.1 * Math.min(f.functions.size + f.refs.size, 6);
  return base * complexity;
}

// Maximum-cardinality bipartite matching via Kuhn's augmenting-path algorithm, restricted to edges
// scoring >= threshold. A plain greedy "sort candidates by score, take first-come-first-served"
// pass under-counts matches: whichever side of a contested pair loses out is left unmatched even
// when a different, equally-valid assignment would have matched EVERY measure that has a
// candidate. Example: A0~B0=0.90, A0~B1=0.85, A1~B0=0.88 (A1~B1 below threshold). Greedy takes
// A0-B0 first (highest score) and then discards both A1~B0 (B0 taken) and A0~B1 (A0 taken),
// leaving A1 unmatched -- even though A0-B1 + A1-B0 matches both sides. Kuhn's algorithm finds
// that second assignment by letting a contested node "steal" its match and pushing the displaced
// node to search for an alternative (an augmenting path), which is guaranteed to find a maximum
// matching regardless of processing order. Node order (by each node's best candidate score,
// descending) and per-node edge order (by score, descending) are both deterministic tie-breakers
// that bias the search toward higher-weight matchings among the (possibly several) maximum ones.
function maxWeightBipartiteMatch(
  scoreOf: (i: number, j: number) => number,
  sizeA: number,
  sizeB: number,
  threshold: number,
): Array<[number, number, number]> {
  const adj: Array<Array<[number, number]>> = []; // adj[i] = [[score, j], ...] desc by score, then j
  for (let i = 0; i < sizeA; i++) {
    const row: Array<[number, number]> = [];
    for (let j = 0; j < sizeB; j++) {
      const score = scoreOf(i, j);
      if (score >= threshold) row.push([score, j]);
    }
    row.sort((x, y) => y[0] - x[0] || x[1] - y[1]);
    adj.push(row);
  }

  const matchB = new Map<number, number>(); // j -> i
  const matchA = new Map<number, number>(); // i -> j
  const matchScore = new Map<number, number>(); // i -> score of its current match

  function tryAugment(i: number, visitedB: Set<number>): boolean {
    for (const [score, j] of adj[i]) {
      if (visitedB.has(j)) continue;
      visitedB.add(j);
      const occupant = matchB.get(j);
      if (occupant === undefined || tryAugment(occupant, visitedB)) {
        matchB.set(j, i);
        matchA.set(i, j);
        matchScore.set(i, score);
        return true;
      }
    }
    return false;
  }

  const order = Array.from({ length: sizeA }, (_, i) => i)
    .filter((i) => adj[i].length > 0)
    .sort((x, y) => adj[y][0][0] - adj[x][0][0] || x - y);
  for (const i of order) tryAugment(i, new Set());

  const result: Array<[number, number, number]> = [];
  for (const [i, j] of matchA) result.push([matchScore.get(i)!, i, j]);
  result.sort((x, y) => y[0] - x[0] || x[1] - y[1] || x[2] - y[2]); // highest-confidence matches first
  return result;
}

export function matchModelMeasures(
  measuresA: Measure[],
  measuresB: Measure[],
  threshold = 0.8,
): MeasureMatch {
  // Drop measures whose DAX is unavailable (empty) so they neither match each other nor dilute the
  // weights — otherwise a tenant that withholds expressions scores every model pair as a clone (acc1).
  const realA = measuresA.filter((m) => (m.dax ?? "").trim() !== "");
  const realB = measuresB.filter((m) => (m.dax ?? "").trim() !== "");
  const featsA = realA.map((m) => extractFeatures(m.dax));
  const featsB = realB.map((m) => extractFeatures(m.dax));
  const weightsA = realA.map((m, i) => measureWeight(m, featsA[i]));
  const weightsB = realB.map((m, i) => measureWeight(m, featsB[i]));

  const assignment = maxWeightBipartiteMatch(
    (i, j) => measureSimilarity(featsA[i], featsB[j]),
    featsA.length,
    featsB.length,
    threshold,
  );

  const matched: Array<{ a: string; b: string; score: number }> = [];
  let matchedWeight = 0;
  let strongMatched = 0;
  for (const [score, i, j] of assignment) {
    matched.push({ a: realA[i].name, b: realB[j].name, score });
    matchedWeight += Math.min(weightsA[i], weightsB[j]) * score;
    // Ref-backed = identical DAX, an exact qualified column match, or a shared SPECIFIC bare name. A
    // pure structural shape match (disjoint refs) or a generic-name-only coincidence is NOT counted,
    // so it can surface for review but never manufactures a strong dup.
    if (isRefBacked(featsA[i], featsB[j])) {
      strongMatched += 1;
    }
  }

  const totalA = weightsA.reduce((s, w) => s + w, 0);
  const totalB = weightsB.reduce((s, w) => s + w, 0);
  const maxT = Math.max(totalA, totalB);
  const minT = Math.min(totalA, totalB);
  const similarity = maxT ? matchedWeight / maxT : 0;
  const containment = minT ? matchedWeight / minT : 0;
  return { similarity: round4(similarity), containment: round4(Math.min(1, containment)), matched, strongMatched };
}

File: semantic_sweep/measures.py
"""Measure-level similarity: weighted lexical DAX features + maximum bipartite matching.

We deliberately avoid embeddings (a general text model collapses ``SUM`` vs ``AVERAGE``). Instead
each measure becomes a small feature set (functions, referenced columns/measures, operators,
context flags, a clone *skeleton*) and pairs are scored by a weighted blend with two guards:
an **incompatible-aggregator penalty** (``SUM`` vs ``AVERAGE``) and **function-family partial
credit** (``ROUND`` vs ``TRUNC``). Models are then matched one-to-one via Kuhn's augmenting-path
algorithm (see ``_max_weight_bipartite_match``), which finds a maximum-cardinality assignment
instead of the smaller matching a naive greedy first-come-first-served pass can get stuck with.
"""

from __future__ import annotations

import math
import re
from collections.abc import Callable
from dataclasses import dataclass, field

from semantic_sweep.parser import Measure

AGGREGATORS = {
    "sum",
    "average",
    "min",
    "max",
    "count",
    "counta",
    "countrows",
    "distinctcount",
    "sumx",
    "averagex",
    "minx",
    "maxx",
    "countx",
    "product",
    "productx",
    "median",
    "geomean",
}
FUNCTION_FAMILIES = {
    "round": "rounding",
    "roundup": "rounding",
    "rounddown": "rounding",
    "trunc": "rounding",
    "int": "rounding",
    "mround": "rounding",
    "fixed": "rounding",
    "ceiling": "rounding",
    "floor": "rounding",
    "sum": "additive",
    "sumx": "additive",
    "average": "mean",
    "averagex": "mean",
    "count": "counting",
    "counta": "counting",
    "countrows": "counting",
    "countx": "counting",
    "distinctcount": "counting",
}
CONTEXT_FLAGS = {
    "calculate",
    "calculatetable",
    "filter",
    "all",
    "allexcept",
    "allselected",
    "removefilters",
    "userelationship",
    "keepfilters",
    "sameperiodlastyear",
    "dateadd",
    "totalytd",
    "totalmtd",
    "totalqtd",
    "datesytd",
    "parallelperiod",
    "previousmonth",
    "previousyear",
    "datesinperiod",
}
GENERIC_NAMES = {
    "total",
    "count",
    "sum",
    "amount",
    "value",
    "measure",
    "result",
    "kpi",
    "average",
    "max",
    "min",
}

# Column/measure ref names common enough to recur across unrelated tables on any estate with a
# shared naming convention (e.g. every fact table has an [Amount], every dimension has a [Date]). A
# bare match on one of these must not, by itself, count as ref-backed strong-duplicate evidence
# (acc6) -- unlike GENERIC_NAMES above (measure-name weighting), this gates measure-level evidence.
GENERIC_REF_NAMES = {
    "id",
    "key",
    "code",
    "name",
    "date",
    "value",
    "amount",
    "total",
    "count",
    "sum",
    "status",
    "type",
    "description",
    "flag",
    "number",
    "quantity",
    "price",
    "region",
    "category",
    "year",
    "month",
    "day",
    "created",
    "modified",
    "updated",
}

_BRACKET = re.compile(r"\[([^\]]+)\]")
_REF = re.compile(r"(?:'[^']*'|\w+)?\[[^\]]*\]")
# Same shape as _REF but captures the qualifier (quoted or bare) separately from the bracket content,
# so a "table.column" ref can be built only when a qualifier is actually present in the source DAX.
_QUALIFIED_REF = re.compile(r"(?:'([^']*)'|(\w+))?\[([^\]]+)\]")
# A DAX string literal, escape-aware: an embedded literal `"` is written as a doubled `""` inside the
# string (e.g. `"He said ""hi"""`). A naive `"[^"]*"` stops at the first inner `"`, truncating the
# literal and leaking the remainder as if it were DAX code (dax-tokenizer-hardening).
_STRING = re.compile(r'"(?:[^"]|"")*"')
# One pass, string-aware: tries a full string literal FIRST at each position, so a "//" or "/*" that
# appears INSIDE a string is consumed as part of the string and never misread as a comment start (the
# previous two-pass comment stripper had no string awareness at all).
_COMMENT_OR_STRING = re.compile(_STRING.pattern + r"|/\*.*?\*/|//[^\n]*", re.DOTALL)
_NUMBER = re.compile(r"\b\d+(?:\.\d+)?\b")
_FUNC = re.compile(r"\b([a-z][a-z0-9]*)\s*\(")
_OPERATOR = re.compile(r"[+\-*/&<>=]")

# Weighted blend over *present* feature components (renormalized when some are empty).
_COMPONENT_WEIGHTS = {"refs": 0.45, "functions": 0.30, "flags": 0.15, "operators": 0.10}
_SKELETON_SCORE = 0.92
# Same structural skeleton but the referenced names differ: interpolate between _SKELETON_FLOOR (a
# pure shape collision, e.g. SUM(Sales[x]) vs SUM(HR[y])) and _SKELETON_SCORE (renamed table, same
# column). Kept >= _REVIEW_MEASURE so a fully-renamed clone still surfaces; the strong-duplicate gate
# is enforced separately via ref-backed evidence (MeasureMatch.strong_matched).
_SKELETON_FLOOR = 0.90
_AGG_PENALTY = 0.6


@dataclass(frozen=True)
class DaxFeatures:  # pylint: disable=too-many-instance-attributes
    """Lexical features extracted from one measure's normalized DAX (cohesive feature bag)."""

    norm: str
    skeleton: str
    functions: frozenset[str]
    refs: frozenset[str]
    # "table.column" refs, populated only when a table qualifier is present in the source DAX (e.g.
    # "sales.amount" from Sales[Amount]). Used to require unambiguous evidence before two measures
    # referencing a common *generic* bare name (e.g. [Amount], [Date]) on different tables count as
    # ref-backed strong-duplicate evidence (acc6).
    qualified_refs: frozenset[str]
    aggregators: frozenset[str]
    operators: frozenset[str]
    flags: frozenset[str]


def normalize_dax(dax: str) -> str:
    """Lower-case the DAX and strip comments (string/escape aware) and redundant whitespace."""
    text = _COMMENT_OR_STRING.sub(lambda m: m.group(0) if m.group(0).startswith('"') else " ", dax)
    return re.sub(r"\s+", " ", text.lower()).strip()


def _skeleton(norm: str) -> str:
    """Structural skeleton: blank refs (incl. table prefix)/strings/numbers so renamed clones collide."""
    text = _REF.sub("#r#", norm)
    text = _STRING.sub("#s#", text)
    text = _NUMBER.sub("#n#", text)
    return re.sub(r"\s+", "", text)


def _extract_qualified_refs(no_strings: str) -> frozenset[str]:
    """Extract "table.column" refs, only when a table qualifier is present in the source DAX."""
    out = set()
    for quoted, bare, name in _QUALIFIED_REF.findall(no_strings):
        qualifier = quoted or bare
        if qualifier:
            out.add(f"{qualifier}.{name}")
    return frozenset(out)


def extract_features(dax: str) -> DaxFeatures:
    """Extract :class:`DaxFeatures` from a raw DAX expression."""
    norm = normalize_dax(dax)
    functions = frozenset(_FUNC.findall(norm))
    # Neutralize string literals before extracting refs so a format string like FORMAT(x, "[Red]0")
    # does not leak a bogus "red" reference (acc3).
    no_strings = _STRING.sub('""', norm)
    return DaxFeatures(
        norm=norm,
        skeleton=_skeleton(norm),
        functions=functions,
        refs=frozenset(m.lower() for m in _BRACKET.findall(no_strings)),
        qualified_refs=_extract_qualified_refs(no_strings),
        aggregators=frozenset(functions & AGGREGATORS),
        operators=frozenset(_OPERATOR.findall(norm)),
        flags=frozenset(functions & CONTEXT_FLAGS),
    )


def _is_ref_backed(a: DaxFeatures, b: DaxFeatures) -> bool:
    """True if two matched measures share genuine reference evidence, not a coincidental collision.

    Identical DAX or an exact table-qualified column/measure match is unambiguous. A shared BARE name
    is only accepted when it is specific enough to not be a coincidence -- a shared *generic* name
    like [Amount] or [Date] recurs across unrelated tables on any estate with a common naming
    convention and must not manufacture strong-duplicate evidence on its own (acc6). Renamed-table
    clones with specific shared column names (e.g. Sales[amt0] ~ Revenue[amt0]) still count,
    preserving the acc5 renamed-clone heuristic.
    """
    if a.norm == b.norm:
        return True
    if a.qualified_refs & b.qualified_refs:
        return True
    shared = a.refs & b.refs
    return any(name not in GENERIC_REF_NAMES for name in shared)


def _jaccard(a: frozenset[str], b: frozenset[str]) -> float:
    if not a and not b:
        return 0.0
    return len(a & b) / len(a | b)


def round4(x: float) -> float:
    """Round to 4 decimals using round-half-up, matching the TS engine's ``Math.round``-based
    ``round4`` (engine/types.ts). Python's builtin ``round()`` uses banker's rounding (ties-to-even),
    which disagrees with JS on exact 4th-decimal ties -- e.g. ``round(0.03125, 4)`` is ``0.0312`` in
    Python but ``0.0313`` in JS. Every value passed here (jaccard ratios, weighted headline scores)
    is non-negative, so round-half-up is unambiguous and reproduces the JS result exactly.
    """
    return math.floor(x * 10000 + 0.5) / 10000


def _function_similarity(a: frozenset[str], b: frozenset[str]) -> float:
    """Jaccard over functions, but same-family functions (round/trunc) get half credit."""
    if not a and not b:
        return 0.0
    exact = a & b
    rest_a = {FUNCTION_FAMILIES[f] for f in (a - exact) if f in FUNCTION_FAMILIES}
    rest_b = {FUNCTION_FAMILIES[f] for f in (b - exact) if f in FUNCTION_FAMILIES}
    soft = len(exact) + 0.5 * len(rest_a & rest_b)
    return min(1.0, soft / len(a | b))


def measure_similarity(a: DaxFeatures, b: DaxFeatures) -> float:
    """Similarity in [0, 1] between two measures' features (1.0 = identical normalized DAX)."""
    # No DAX on either side (e.g. a locked-down / admin scan that withholds expressions) is NOT
    # evidence of a match -- two shape-identical models must not read as exact clones (acc1).
    if not a.norm and not b.norm:
        return 0.0
    if a.norm == b.norm:
        return 1.0
    if a.skeleton and a.skeleton == b.skeleton:
        return round4(_SKELETON_FLOOR + (_SKELETON_SCORE - _SKELETON_FLOOR) * _jaccard(a.refs, b.refs))
    scores = {
        "refs": _jaccard(a.refs, b.refs),
        "functions": _function_similarity(a.functions, b.functions),
        "flags": _jaccard(a.flags, b.flags),
        "operators": _jaccard(a.operators, b.operators),
    }
    present = {
        ("refs", a.refs or b.refs),
        ("functions", a.functions or b.functions),
        ("flags", a.flags or b.flags),
        ("operators", a.operators or b.operators),
    }
    active = [(k, _COMPONENT_WEIGHTS[k]) for k, has in present if has]
    if not active:
        return 0.0
    base = sum(w * scores[k] for k, w in active) / sum(w for _, w in active)
    if a.aggregators and b.aggregators and a.aggregators.isdisjoint(b.aggregators) and (a.refs & b.refs):
        base *= _AGG_PENALTY  # SUM(x) vs AVERAGE(x): same column, incompatible aggregator
    return round4(min(1.0, base))


def _measure_weight(measure: Measure, feats: DaxFeatures) -> float:
    """Informativeness weight: generic-named measures count less; complex ones count more.

    Tokenize on word boundaries (not a single concatenated blob) so a multi-word name like "Total
    Sales" is judged word-by-word against GENERIC_NAMES instead of vanishing into "totalsales",
    which matches nothing and was silently never downweighted. Square the generic-word fraction
    before applying the discount: a name that is ALL generic words (e.g. "Total", or "Total Count")
    still gets the full 0.4 floor exactly as before, but a name that is only PARTLY generic (e.g.
    "Total Sales") is discounted much more gently, since one specific word ("Sales") already carries
    most of the discriminating signal -- informativeness doesn't fall off linearly with word count.
    """
    words = [w for w in re.split(r"[^a-z0-9]+", measure.name.lower()) if w]
    generic_frac = (sum(1 for w in words if w in GENERIC_NAMES) / len(words)) if words else 0.0
    base = 1.0 - 0.6 * generic_frac * generic_frac
    complexity = 1.0 + 0.1 * min(len(feats.functions) + len(feats.refs), 6)
    return base * complexity


@dataclass
class MeasureMatch:
    """Result of matching two models' measure sets."""

    similarity: float
    containment: float
    matched: list[tuple[str, str, float]] = field(default_factory=list)
    # Matched pairs backed by real reference overlap (identical DAX or a shared column), not just a
    # coincidental structural shape. Gates the strong-duplicate / subset bands (see score.py).
    strong_matched: int = 0


def _max_weight_bipartite_match(
    score_of: Callable[[int, int], float], size_a: int, size_b: int, threshold: float
) -> list[tuple[float, int, int]]:
    """Maximum-cardinality bipartite matching via Kuhn's augmenting-path algorithm.

    A plain greedy "sort candidates by score, take first-come-first-served" pass under-counts
    matches: whichever side of a contested pair loses out is left unmatched even when a different,
    equally-valid assignment would have matched EVERY measure that has a candidate. Example:
    A0~B0=0.90, A0~B1=0.85, A1~B0=0.88 (A1~B1 below threshold). Greedy takes A0-B0 first (highest
    score) and then discards both A1~B0 (B0 taken) and A0~B1 (A0 taken), leaving A1 unmatched --
    even though A0-B1 + A1-B0 matches both sides. Kuhn's algorithm finds that second assignment by
    letting a contested node "steal" its match and pushing the displaced node to search for an
    alternative (an augmenting path), which is guaranteed to find a maximum matching regardless of
    processing order. Node order (by each node's best candidate score, descending) and per-node
    edge order (by score, descending) are both deterministic tie-breakers that bias the search
    toward higher-weight matchings among the (possibly several) maximum ones.
    """
    adj: list[list[tuple[float, int]]] = []  # adj[i] = [(score, j), ...] desc by score, then j
    for i in range(size_a):
        row = [(score_of(i, j), j) for j in range(size_b) if score_of(i, j) >= threshold]
        row.sort(key=lambda t: (-t[0], t[1]))
        adj.append(row)

    match_b: dict[int, int] = {}  # j -> i
    match_a: dict[int, int] = {}  # i -> j
    match_score: dict[int, float] = {}  # i -> score of its current match

    def try_augment(i: int, visited_b: set[int]) -> bool:
        for score, j in adj[i]:
            if j in visited_b:
                continue
            visited_b.add(j)
            occupant = match_b.get(j)
            if occupant is None or try_augment(occupant, visited_b):
                match_b[j] = i
                match_a[i] = j
                match_score[i] = score
                return True
        return False

    order = sorted((i for i in range(size_a) if adj[i]), key=lambda i: (-adj[i][0][0], i))
    for i in order:
        try_augment(i, set())

    result = [(match_score[i], i, j) for i, j in match_a.items()]
    result.sort(key=lambda t: (-t[0], t[1], t[2]))  # highest-confidence matches first
    return result


def match_model_measures(
    measures_a: list[Measure], measures_b: list[Measure], *, threshold: float = 0.8
) -> MeasureMatch:
    """Maximum one-to-one match between two measure sets; return similarity + containment.

    ``similarity = matched_weight / max(total)`` (penalizes size gaps), while
    ``containment = matched_weight / min(total)`` flags a subset/trimmed copy.
    """
    # pylint: disable=too-many-locals  # cohesive matching algorithm
    # Drop measures whose DAX is unavailable (empty) so they neither match each other nor dilute the
    # weights -- otherwise a tenant that withholds expressions scores every model pair as a clone (acc1).
    real_a = [m for m in measures_a if (m.dax or "").strip()]
    real_b = [m for m in measures_b if (m.dax or "").strip()]
    feats_a = [extract_features(m.dax) for m in real_a]
    feats_b = [extract_features(m.dax) for m in real_b]
    weights_a = [_measure_weight(m, f) for m, f in zip(real_a, feats_a)]
    weights_b = [_measure_weight(m, f) for m, f in zip(real_b, feats_b)]

    assignment = _max_weight_bipartite_match(
        lambda i, j: measure_similarity(feats_a[i], feats_b[j]), len(feats_a), len(feats_b), threshold
    )

    matched: list[tuple[str, str, float]] = []
    matched_weight = 0.0
    strong_matched = 0
    for score, i, j in assignment:
        matched.append((real_a[i].name, real_b[j].name, score))
        matched_weight += min(weights_a[i], weights_b[j]) * score
        # Ref-backed = identical DAX, an exact qualified column match, or a shared SPECIFIC bare name.
        # A pure structural shape match (disjoint refs) or a generic-name-only coincidence is not
        # counted, so it surfaces for review but never manufactures a dup.
        if _is_ref_backed(feats_a[i], feats_b[j]):
            strong_matched += 1

    total_a, total_b = sum(weights_a), sum(weights_b)
    similarity = matched_weight / max(total_a, total_b) if max(total_a, total_b) else 0.0
    containment = matched_weight / min(total_a, total_b) if min(total_a, total_b) else 0.0
    return MeasureMatch(
        similarity=round4(similarity),
        containment=round4(min(1.0, containment)),
        matched=matched,
        strong_matched=strong_matched,
    )
